### PR TITLE
 chore: automate aurora db engine updates in GH Actions

### PR DESCRIPTION
**Pull request by jasonlamz**
### Summary

Automate Aurora integration tests to use both latest and LTS versions

### Description

- Modify integration tests to use LTS and latest versions
- Allow user specified system property to override versions
- Use SDK to get db engine versions

### By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/run-integration-tests.yml [hunk 1]
```diff
@@ -11,8 +11,8 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-  all-integration-tests:
-    name: 'Run Aurora integration tests'
+  all-integration-tests-latest:
```
Comment (sergiyvamz) [inline .github/workflows/run-integration-tests.yml:14]: Integration may run for hours and sometimes it reach the max allowed run time of 6h. Increasing  (doubling) amount of work for this github action will definitely increase the chance of such timeout out event. I'd suggest to split this script into two separate github actions.

File: wrapper/src/test/java/integration/host/TestEnvironment.java [hunk 2]
```diff
@@ -471,15 +473,38 @@ private static String getAuroraDbEngine(TestEnvironmentRequest request) {
...
     }
+    return findAuroraDbEngineVersion(env, engineName, systemPropertyVersion);
+  }
+
+  private static String findAuroraDbEngineVersion(
+      TestEnvironment env,
+      String engineName,
+      String systemPropertyVersion
+  ) {
+    if (systemPropertyVersion == null || systemPropertyVersion.equals("lts")) {
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironment.java:501]: Can we use switch-case syntax for readability?

File: wrapper/src/test/java/integration/util/AuroraTestUtility.java [hunk 3]
```diff
@@ -923,4 +927,26 @@ public void addAuroraAwsIamUser(
...
+        DescribeDbEngineVersionsRequest.builder().engine(engine).build()
+    );
+    for (DBEngineVersion version : versions.dbEngineVersions()) {
+      res.add(version.engineVersion());
+    }
+    return res;
+  }
+
+  public String getLatestVersion(String engine) {
+    return Collections.max(getEngineVersions(engine));
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/util/AuroraTestUtility.java:943]: Why "max"? Maybe collection.stream().sorted().last() ?

File: wrapper/src/test/java/integration/util/AuroraTestUtility.java [hunk 4]
```diff
@@ -923,4 +927,26 @@ public void addAuroraAwsIamUser(
...
+  }
+
+  public String getLTSVersion(String engine) {
+    final DescribeDbEngineVersionsResponse versions = rdsClient.describeDBEngineVersions(
+        DescribeDbEngineVersionsRequest.builder().defaultOnly(true).engine(engine).build()
+    );
+    return versions.dbEngineVersions().get(0).engineVersion();
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/util/AuroraTestUtility.java:950]: Can we check if collection isn't empty?

File: wrapper/src/test/java/integration/host/TestEnvironment.java [hunk 5]
```diff
@@ -471,15 +473,38 @@ private static String getAuroraDbEngine(TestEnvironmentRequest request) {
     }
   }
 
-  private static String getAuroraDbEngineVersion(TestEnvironmentRequest request) {
+  private static String getAuroraDbEngineVersion(
+      TestEnvironment env,
+      TestEnvironmentRequest request) {
+    String engineName;
+    String systemPropertyVersion;
     switch (request.getDatabaseEngine()) {
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironment.java:481]: We need support of TAZ clusters either

File: wrapper/src/test/java/integration/host/TestEnvironmentConfiguration.java [hunk 6]
```diff
@@ -78,6 +78,10 @@ public class TestEnvironmentConfiguration {
   public String auroraClusterName = System.getenv("AURORA_CLUSTER_NAME"); // "cluster-mysql"
   public String auroraClusterDomain =
       System.getenv("AURORA_CLUSTER_DOMAIN"); // "XYZ.us-west-2.rds.amazonaws.com"
+  public String auroraMySqlDbEngineVersion =
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironmentConfiguration.java:81]: Can we add a comment with a possible/expected values?

File: .github/workflows/run-integration-tests.yml (hunk 1 above)
Comment (karenc-bq) [reply .github/workflows/run-integration-tests.yml:14]: Adding to what @sergiyvamz said, another way to run multiple jobs in parallel without duplicating code is to use matrix https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs

File: wrapper/src/test/java/integration/host/TestEnvironment.java [hunk 7]
```diff
@@ -471,15 +473,41 @@ private static String getAuroraDbEngine(TestEnvironmentRequest request) {
     }
   }
 
-  private static String getAuroraDbEngineVersion(TestEnvironmentRequest request) {
+  private static String getAuroraDbEngineVersion(
+      TestEnvironment env,
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironment.java:477]: "env" contains all that the method needs. I think a single parameter would be enough.

File: wrapper/src/test/java/integration/host/TestEnvironment.java [hunk 8]
```diff
@@ -471,15 +473,41 @@ private static String getAuroraDbEngine(TestEnvironmentRequest request) {
...
+  }
+
+  private static String findAuroraDbEngineVersion(
+      TestEnvironment env,
+      String engineName,
+      String systemPropertyVersion) {
+    if (systemPropertyVersion == null) {
+      return env.auroraUtil.getLTSVersion(engineName);
+    }
+    switch (systemPropertyVersion) {
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironment.java:503]: systemPropertyVersion.toLower()

File: wrapper/src/test/java/integration/util/AuroraTestUtility.java [hunk 9]
```diff
@@ -923,4 +928,33 @@ public void addAuroraAwsIamUser(
...
+      res.add(version.engineVersion());
+    }
+    return res;
+  }
+
+  public String getLatestVersion(String engine) {
+    return getEngineVersions(engine)
+        .stream()
+        .sorted()
+        .reduce((first, last) -> last)
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/util/AuroraTestUtility.java:947]: Can we use sorted() to sort collection and then return first() or last()?
https://www.baeldung.com/java-8-sort-lambda

File: wrapper/src/test/java/integration/host/TestEnvironment.java [hunk 10]
```diff
@@ -390,9 +389,12 @@ private static void createAuroraDbCluster(TestEnvironment env, int numOfInstance
 
       try {
         String engine = getAuroraDbEngine(env.info.getRequest());
-        String engineVersion = getAuroraDbEngineVersion(env.info.getRequest());
+        String engineVersion = getAuroraDbEngineVersion(env);
```
Comment (sergiyvamz) [inline wrapper/src/test/java/integration/host/TestEnvironment.java:392]: getAuroraDbEngineVersion() may return null. We need to check it.

**Review by sergiyvamz — APPROVED** [review]

**Pull request merged**